test/boost/commitlog_test.cc

```diff
@@ -1822,3 +1822,205 @@ SEASTAR_TEST_CASE(test_commitlog_update_max_data_lifetime) {
+            tot += mutations.back().representation().size();
+        }
+        for (auto& fm : mutations) {
+            writers.emplace_back(gen.schema(), fm, commitlog::force_sync::no);
+        }
+
+        // this will create an oversized entry set.
+        auto res = co_await log.add_entries(writers, db::timeout_clock::now() + 60s);
+
+        auto i = mutations.begin();
+        for (auto& h : res) {
+            rp2mut.emplace(h.release(), *i++);
+        }
+        co_await log.sync_all_segments();
+        // as if we crashed -> segment left on disk
+        co_await log.release();
+        co_await log.shutdown();
+    }
+
+    // new log, for replay.
+    auto log = co_await commitlog::create_commitlog(cfg);
+    std::exception_ptr e;
+    size_t n = 0;
+
+    auto replay_set = co_await log.get_segments_to_replay();
+    // Now replay the old commitlog and ensure we match all data.
+
+    commitlog::replay_state state;
+    for (auto& f : replay_set) {
+        try {
+            co_await commitlog::read_log_file(state, f, cfg.fname_prefix, [&](commitlog::buffer_and_replay_position buf_rp) -> future<> {
+                auto&& buf = buf_rp.buffer;
+                auto&& rp = buf_rp.position;
+
+                BOOST_CHECK(rp2mut.count(rp));
+                commitlog_entry_reader cer(buf);
+                auto& fm = cer.mutation();
+                auto m1 = fm.unfreeze(gen.schema());
+                auto m2 = rp2mut.at(rp).unfreeze(gen.schema());
+
+                BOOST_CHECK_EQUAL(m1, m2);
+                ++n;
+                co_return;
+            });
+        } catch (commitlog::segment_truncation&) {
+            e = std::current_exception();
+        }
+    }
+
+    BOOST_CHECK_EQUAL(n, rp2mut.size());
+
+    co_await log.shutdown();
+    co_await log.clear();
+
+    if (n != rp2mut.size() && e) {
+        std::rethrow_exception(e);
+    }
+}
+
+SEASTAR_TEST_CASE(test_oversized_entry_small) {
+    co_await do_test_oversized_entry(1); // small segments
+}
+
+SEASTAR_TEST_CASE(test_oversized_entry_normal) {
+    co_await do_test_oversized_entry(32); // normal segments
+}
+
+SEASTAR_TEST_CASE(test_oversized_entry_large) {
+    co_await do_test_oversized_entry(32*3); // bigger segments
+}
+
+static future<> test_oversized(size_t n_entries, size_t max_size_mb) {
+    commitlog::config cfg;
+
+    cfg.commitlog_segment_size_in_mb = max_size_mb;
+    cfg.commitlog_total_space_in_mb = 8 * n_entries * max_size_mb * smp::count;
+    cfg.allow_going_over_size_limit = false;
+    cfg.allow_fragmented_entries = true;
+    cfg.use_o_dsync = false; 
+
+    // not using cl_test, because we need to be able to abandon
+    // the log.
+    tmpdir tmp;
+    cfg.commit_log_location = tmp.path().string();
+
+    auto uuid = make_table_id();
+    std::unordered_map<replay_position, fragmented_temporary_buffer> rp2buf;
+    {
+        std::random_device rd;
+        std::mt19937 gen(rd());
+        std::uniform_int_distribution<char> dist;
```

Comment:
>  test/boost/commitlog_test.cc:1949:49: error: taking the absolute value of unsigned type 'result_type' (aka 'char') has no effect [-Werror,-Wabsolute-value]
> 00:48:19   1949 |         auto size = log.max_record_size() * 2 + std::abs(dist(gen)) * 1024 + std::abs(dist(gen)) * 64;
> 00:48:19        |                                                 ^
> 00:48:19  test/boost/commitlog_test.cc:1949:49: note: remove the call to 'abs' since unsigned values cannot be negative

To my understanding, [you can't use `char`](https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution) as a template parameter to [uniform_int_distribution](https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution) - the behavior is undefined.

```suggestion
std::uniform_int_distribution<short> dist;
```

Reply:
Strange that we only hit this on aarch64. I expect such errors to be platform independent. Or maybe we have different default char signedness on amd64 vs. aarch64?

Reply:
Note we use distributions of char on several places - they should be fine. Weird with char being unsigned on arm though.

Reply:
Here we are not even using char values - just want 0-256 -> std::abs(int8_t) is fine.

Reply:
Or even simpler, just unsigned 0>255

Reply:
> Strange that we only hit this on aarch64

Strictly speaking, UB (Undefined Behavior) can result in _anything_ - from the application moving your bank account to me, executing `rm -rf /` and also, sometimes, the expected behavior. :-)

Reply:
This isn't a UBSAN error in runtime, but rather a compile-time error in:
```
        auto size = log.max_record_size() * 2 + std::abs(dist(gen)) * 1024 + std::abs(dist(gen)) * 64;
```

where apparently `dist(gen)` for `std::uniform_int_distribution<char> dist` results in unsigned char unexpectedly.

Reply:
I've  changed it to simply just use an unsigned and restrict the bit width. But we use the pattern in more than one  test.